File: src/main/java/raft/war/binary/parser/w3g/parser/commandblock/actions/ujapi/UjApiKeyAction.java
package raft.war.binary.parser.w3g.parser.commandblock.actions.ujapi;

import java.nio.ByteBuffer;

public class UjApiKeyAction implements IUjapiSubAction {

    private final byte subActionId;

    private short key;

    private byte metaKey;

    private boolean isKeyDown;

    public UjApiKeyAction(byte subActionId) {
        this.subActionId = subActionId;
    }

    @Override
    public void parse(ByteBuffer inBuffer) {
        this.key = inBuffer.getShort();
        this.metaKey = inBuffer.get();
        this.isKeyDown = inBuffer.get() > 0;
    }

    @Override
    public ByteBuffer assembly(ByteBuffer outBuffer) {
        if (outBuffer == null) {
            outBuffer = ByteBuffer.allocate(4).order(java.nio.ByteOrder.LITTLE_ENDIAN);
        }

        outBuffer.putShort(key);
        outBuffer.put(metaKey);
        outBuffer.put((byte) (isKeyDown ? 1 : 0));

        return outBuffer;
    }

    public short getKey() {
        return key;
    }

    public UjApiKeyAction setKey(short key) {
        this.key = key;
        return this;
    }

    public byte getMetaKey() {
        return metaKey;
    }

    public UjApiKeyAction setMetaKey(byte metaKey) {
        this.metaKey = metaKey;
        return this;
    }

    public boolean isKeyDown() {
        return isKeyDown;
    }

    public UjApiKeyAction setKeyDown(boolean keyDown) {
        isKeyDown = keyDown;
        return this;
    }

    @Override
    public byte getSubActionId() {
        return subActionId;
    }
}

File: src/main/kotlin/raft/war/binary/parser/w3g/commandBlock/action/ujapi/UjapiKeyAction.kt
package raft.war.binary.parser.w3g.commandBlock.action.ujapi

import java.nio.ByteBuffer

class UjapiKeyAction(override val subActionId: Byte) : UjapiSubActionBase {
    private var key: Short = 0
    private var metaKey: Byte = 0
    private var isKeyDown = false

    override fun parse(inBuffer: ByteBuffer?) {
        this.key = inBuffer!!.getShort()
        this.metaKey = inBuffer.get()
        this.isKeyDown = inBuffer.get() > 0
    }

    override fun toString(): String {
        return "UjapiKeyAction\n" +
                "\t\t\tkey=$key\n" +
                "\t\t\tmetaKey=$metaKey\n" +
                "\t\t\tisKeyDown=$isKeyDown\n"
    }
}
